src/ring_theory/finiteness.lean

```diff
@@ -550,3 +551,55 @@ ring_hom.finite_presentation.of_finite_type
 end finite_presentation
 
 end alg_hom
+
+section monoid_algebra
+
+namespace add_monoid_algebra
+
+open algebra submodule
+
+variables {R : Type*} {M : Type*} [comm_ring R] [add_comm_monoid M]
+
+instance ft_of_fg [h : add_monoid.fg M] : finite_type R (add_monoid_algebra R M) :=
+begin
+  obtain ⟨S, hS⟩ := h.out,
+  let S₁ := (S : set M),
+  let ψ := λ (s : S₁), of R M (multiplicative.of_add ↑s),
+  apply (finite_type.mv_polynomial R S₁).of_surjective (mv_polynomial.aeval ψ),
+  refine λ f, finsupp.induction_linear f _ _ _,
+  { exact ⟨0, alg_hom.map_zero _⟩ },
+  { rintro f g ⟨P, rfl⟩ ⟨Q, rfl⟩,
+    exact ⟨P + Q, alg_hom.map_add _ _ _⟩, },
+{ intros m r,
+    suffices : ∃ (P : mv_polynomial S₁ R), (mv_polynomial.aeval ψ) P = finsupp.single m 1,
+    { obtain ⟨P, hP⟩ := this,
+      exact ⟨r • P, by simp only [hP, alg_hom.map_smul, mul_one, finsupp.smul_single']⟩ },
+    have : m ∈ span ℕ (S : set M),
+    { rw [← mem_to_add_submonoid, span_nat_eq_add_submonoid_closure, hS],
+      exact add_submonoid.mem_top _, },
+    refine span_induction this _ _ _ _,
+    { intros m hm,
+      use mv_polynomial.X ⟨m, hm⟩,
+      simp only [ψ, mv_polynomial.aeval_X, of_apply, subtype.coe_mk],
+      refl },
+    { exact ⟨1, alg_hom.map_one _⟩ },
+    { rintro m₁ m₂ ⟨P₁, hP₁⟩ ⟨P₂, hP₂⟩,
+      exact ⟨P₁ * P₂, by rw [alg_hom.map_mul, hP₁, hP₂, single_mul_single, one_mul]⟩ },
+    { rintro a m ⟨P, hP⟩,
+      exact ⟨P ^ a, by rw [alg_hom.map_pow, hP, single_pow, one_pow]⟩ } },
```

Comment:
`span` is a distraction here (but your `mem_to_add_submonoid` lemma is worth keeping in the PR anyway)
```suggestion
    have : m ∈ add_submonoid.closure (S : set M) := hS.symm ▸ add_submonoid.mem_top _,
    refine add_submonoid.closure_induction this _ _ _,
    { intros m hm,
      use mv_polynomial.X ⟨m, hm⟩,
      simp only [ψ, mv_polynomial.aeval_X, of_apply, subtype.coe_mk],
      refl },
    { exact ⟨1, alg_hom.map_one _⟩ },
    { rintro m₁ m₂ ⟨P₁, hP₁⟩ ⟨P₂, hP₂⟩,
      exact ⟨P₁ * P₂, by rw [alg_hom.map_mul, hP₁, hP₂, single_mul_single, one_mul]⟩ }, },
```